### ISSUE TITLE
support for RC4 known-plaintext attack

### ISSUE THREAD

**Issue by philsmd** [hashcat/oclHashcat#88]
This feature request was originally reported on the (now obsolete and offline) **trac** ticketing system of hashcat.net.

---

| Ticket details:                    |               |
| -------------------------------- | ---------- |
| Original reporter (OP):      | RC4_Support |
| Title:                                  | RC4 known-plaintext attack support |
| Ticket number:                  | 596 |
| Date reported:                  | 2015-03-03T08:36:01+01:00 |
| **Description**:                 |  [The i]dea [here] is [to implement a] kernel for [a] very generic RC4 plainttext attack.<br><br>We already have some RC4 modules (pdf/office) so we can modify it. How it works? [The u]ser input[s] known bytes before, crypt bytes after (with same length). User set mask with his knowledge about key and that's all.<br><br>Kernel functionality:<br>- create RC4 initial state from the custom mask<br>- decrypt specified user input with our state (result of step 1)<br>- compare result bytes with specified user known-plaintext (if they match, log result, continue, there can be more keys)<br>- repeat<br><br>[The i]nput is in hex format. Maybe support for mask comparing? e.g<br>Input: 11 22 33 44<br>Output 11 22 ?? 44 (?? mean[s] that it can be any char, 11 22 55 44 is [for instance a] good result) |

**Comment by mattames** [comment]
This would be very handy for myself as well!

**Comment by jdek** [comment]
I'd like to look into this, how would I go about adding a module?

**Comment by bigfella237** [comment]
+1 for me (and the other 160+ watchers), this would get a lot of use here!

RC4 being a stream cipher utilizes variable key lengths with masking, as suggested in the original ticket, but the authors also suggested discarding the first portion of the generated keystream to avoid an inherent weakness. 

This is what's known as "DropN" and is widely implemented in RC4 applications, so the Hashcat version would also need the user to specify the length of the keystream to be generated and the number of bytes to drop from the beginning before the XOR with the plaintext.

In summary, we would input:
-     start key (with mask)
-     key range (number of keys to search)
-     keystream length*
-     DropN length
-     the "target" keystream portion (to compare with the keystream generated by each pass)

*Alternately the keystream length could easily be calculated automatic by adding the DropN value entered to the Target length.

The Target keystream is derived from a known plaintext XOR the ciphertext. Once our Target matches the generated keystream we've found the key!

Please, someone, make my day and implement this in Hashcat?

**Comment by jsteube** [comment]
I've recently added KPA for DES. I'm thinking about doing the same for RC4 after the refactoring, but I'm thinking about a fixed 40 bit key length, not a variable one. Also I don't think about adding support for that DropN length, except if you would know a software or protocol that uses RC4 in 40 bit mode using this technique. The key range can be limited with -l (and offset with -s to allow distributed search).

**Comment by bigfella237** [comment]
My applications require either a 72-bit or 104-bit key length, the last 32 or 64 bits (respectively) of which are known but still must be included in the key, so a "fixed" 40 bit key length would be of no use.

Likewise with DropN, I need to drop either the first 256 or 267 bits (depending on the application) of keystream to align the ciphertext with the correct portion of the stream that was originally used for the XOR.

The beauty of stream ciphers is that they are very flexible, allowing for varying key and output lengths, unlike a block cipher, which has fixed key and IV lengths. With so many different applications of RC4 it seems a pity to restrict Hashcat with a fixed key length.

Is there any advantage to having a fixed key length?

**Comment by mattames** [comment]
Yep. From what I understand, the GPU kernel with a fixed key length can be optimised thus making it faster, since it's not wasting clock cycles dynamically shifting things around...

**Comment by 7805** [comment]
I have several applications similar to what bigfella237 describes. Many systems using RC4 (arc4) have the drop 256 bits before going on.  Ideally it would be nice to be able to set the DropN value as well as the key length.

**Comment by jdek** [comment]
You could always have a few kernels, one 'dynamic' one, and then a few more-optimised ones for common combinations.

**Comment by bigfella237** [comment]
@ joshdekock That is a great idea!

Our project is revisiting some of the work done in the NICTA paper on _"Insecurity in Public-Safety Communications: APCO Project 25"_ (found here: www.nicta.com.au/pub?doc=5076 ) but also extending the scope to include some communications platforms that weren't covered in the original research.

We have three separate platforms that utilize the RC4 algorithm, all have a 40-bit keyspace but also have a known "Message Indicator" concatenated to the key to pseudo-randomize the output. The length of this message indicator varies from one platform to the next.

Likewise with the DropN value, obviously we need to XOR a known plaintext with a ciphertext to get a "target" keystream to test against, but we also need to align the result with the exact portion of the RC4 keystream used on the original XOR. All of these platforms drop the first 256 bytes of keystream, but one has an extra 11 bytes of data sent before our known plaintext that we need to ignore.

So in summary, we would need three variants to cover all three platforms, being:
- 72-bit key with a DropN of 256 bytes;
- 104-bit key with a DropN of 256 bytes;
- 104-bit key with a DropN of 267 bytes.

As stated, all would still only have a 40-bit keyspace to search, but we do need to append the known message indicator as it forms part of the key.

**Comment by mattames** [comment]
Bigfella,

Even though the effective search length between separate searches is 40 bits, the kernel would have to be written as such to accommodate the entire Key+IV as just a raw key, as this is how the cipher is really configured.

With one of the first known RC4 security issues being a weak key-schedule, many protocols that use RC4 as a base algorithm work around this weakness by taking a 40-bit "key" and concatenating this with a separate IV, (for example 64-bits in the case of the NICTA P25 paper) to give a 104-bit effective key length.

Of course, writing many different kernels for obscure modes like this was never the intent of Hashcat, but it would be great to know and demonstrate, as when this was originally researched in 2010, we tested the RC4 on a large Nvidia Tesla cluster, and it was severely limited by the early architecture's lack of shared RAM.

If this could be taken into consideration, that would be greatly appreciated. The recent work in adding DES-ECB has been of great benefit as a base algo, and RC4 would be the cherry on top :)

Cheers,
Matt

**Comment by 7805** [comment]
yes that is along the lines of what I was thinking of doing. it should
be a simple thing to edit the code to accommodate the specifics of the
RC4 system being looked at. Most are WIFI related so the actual setting
are known.

**Comment by bigfella237** [comment]
Sorry if this is a "newb" question (and sorry for my impatience) but is there a way to tell if anyone is working on support for this?

As above, Hashcat already has some RC4 modules in place for other searches and, although I'm no programmer, it shouldn't be too much work to adapt something to suit our needs?

**Comment by bigfella237** [comment]
@philsmd any chance you could take a look at implementing these RC4 modules for us (in your spare time of course)?

I'm guessing there's at least four or five different different "modes" involved depending on whether you can make the key length and/or DropN lengths variable:

- 40-bit key with no DropN;
- 40-bit key with a DropN of 256 bytes;
- 72-bit key with a DropN of 256 bytes;
- 104-bit key with a DropN of 256 bytes;
- 104-bit key with a DropN of 267 bytes.

Any help would be greatly appreciated!

**Comment by magnumripper** [comment]
We could establish some input syntax right now, for a starter. Something like
```
$RC4$<key_len>*<dropN>*<key in hex>*<ciphertext in hex>*<offset of plain>*<plain in hex>

eg.

$RC4$40*256*deadcafe05*31337cafebabeccbbaa*00*08090a0b
```
Would that suffice? Do we need some more field(s)?

**Comment by bigfella237** [comment]
We would just need to add a key mask as we only want to search the first 40 bits (the balance of the key is known), so something like:

hashcat64.exe -m [mode] input.txt -o output.txt --outfile-format 5 -a 3 -1 charset --hex.charset ?1?1?1?1?1knownportion

The above assumes we have a different [mode] for the different variable and DropN lengths, otherwise, as you suggest, those values would also need to be specified in the command line.

**Comment by bigfella237** [comment]
Bumping this request for RC4 known-plaintext attack, it's coming up on 2 years old!

If anyone has the time this would still be very useful to us and shouldn't take long as RC4 is already used in other modes. I wish I knew how to add this myself but I just don't have the skills.

Anyone?

**Comment by bigfella237** [comment]
Just checking-in to wish this request a happy birthday... three years old today!

I would still get a lot of use out of these RC4 known-plaintext attacks, and I'm sure that what would take a CPU several weeks could be done in minutes with Hashcat. 

If anyone has some spare time to look at this I would be eternally grateful, pretty please (with sugar on top)?

**Comment by yuhong** [comment]
Do note that the proposed syntax won't work with all protocols. SSL for example requires an extra MD5 before the key can be tried.

**Comment by bitwisebill** [comment]
@bigfella237 I messaged you on another site about this but I think they nuked my account for looking like a spammer. I think I can help you out, shoot me an email at bitwisebill@protonmail.com if you could please. Thanks!

**Comment by bigfella237** [comment]
I was reviewing this request again as it's just about to turn 5 years old in January and I realized that I never added any verified test results in case anyone is able to have a look at adding the modules.

The first 40-bits of the secret key is what we're searching for (this is unknown in the real world but listed below for this test).

```
For a 72-bit key with a DropN of 256 bytes:
secret first 40-bits of the key: 0x3741586942
known last 32-bits of the key: 0x06069DCC
after dropping the first 256 bytes of keystream should produce the string: 58 2B 45 52 EB D9 7C
```

```
For a 104-bit key with a DropN of 256 bytes:
secret first 40-bits of the key: 0x5974281534
known last 64-bits of the key: 0x5560F28DC480C689
after dropping the first 256 bytes of keystream should produce the string: DB 81 25 F8 3E 69 62
```

```
For a 104-bit key with a DropN of 267 bytes:
secret first 40-bits of the key: 0x9876543210
known last 64-bits of the key: 0x78AD2D347D1612CC
after dropping the first 267 bytes of keystream should produce the string: C5 C5 93 CB 6A C6 F4 19 A3 44 1A
```

**Comment by mattames** [comment]
> @bigfella237 I messaged you on another site about this but I think they nuked my account for looking like a spammer. I think I can help you out, shoot me an email at [bitwisebill@protonmail.com](mailto:bitwisebill@protonmail.com) if you could please. Thanks!

That was over at AusTech right? I think I saw that post...

**Comment by bigfella237** [comment]
And several other forums I frequent!

**Comment by zsneese** [comment]
Has any progress been made on this? I am also attempting exactly what you are attempting and recreating and expanding what is found in that paper. I would love to help out in any way I can.

**Comment by catchxray** [comment]
I have been following this thread closely for a while now as well as other forums. I am no dev however I would be willing to help however I can. I know by the accounts following this thread that published experts in here have been able to defeat RC4 without hashcat I assume however it would make key recovery much easier. I hope someone can implement this new feature set for the community. I can't wait to see what the future holds.

**Comment by michaeldim1** [comment]
I am also interested in the RC4 known-plaintext key search. It would be cool for hashcat to support this for efficiency's sake. I have been _trying_ to write something to do a key search for some time now, and as I know next to nothing about writing code, I have pretty much hit a dead end. I know the broader topic of defeating RC4 outside than hashcat is outside the topic of this thread. If anyone can help me out even a bit please reach out to me at maxcooljazz@protonmail.com

**Comment by N8BWert** [comment]
Not sure if anyone is still wanting of this feature and I have no experience with openCL, but I wrote a multithreaded program in rust that can bruteforce rc4 with a known plaintext.  The link to it is https://github.com/N8BWert/rc4-plaintext.  It has support for known key masks and dropped bytes, which I saw were in demand in this thread.

**Comment by bigfella237** [comment]
Yes, myself and many others I've spoken with would still get a lot of use out of this feature, only it's been almost seven years since it was first requested and despite (what I consider) a reasonable amount of interest it seems nobody is willing to take up the cause within Hashcat. 

It's not the ability to search that's the problem, it's the speed. I'm just guessing here but I would've thought Hashcat could do in mere minutes using Cuda cores what takes weeks on a CPU alone, it's only a 40-bit keyspace after all?

Still hoping for a solution within Hashcat, just not holding my breath!

**Comment by zsneese** [comment]
> Not sure if anyone is still wanting of this feature and I have no experience with openCL, but I wrote a multithreaded program in rust that can bruteforce rc4 with a known plaintext. The link to it is https://github.com/N8BWert/rc4-plaintext. It has support for known key masks and dropped bytes, which I saw were in demand in this thread.

Thank you for being willing to take on the project and produce what most of us haven’t been able to!
I have taken a look at your work and it looks great, just a few concerns.
 It looks as if it’s possible to enter in the known plaintext but no ciphertext. How does the program determine a key using only half of the needed data?

**Comment by michaeldim1** [comment]
I also took a look at it and came away confused to how it works. It asks for a keymask and the known plaintext but not the cyphertext? I presume there's something I'm missing.

**Comment by N8BWert** [comment]
I should probably be more clear in the readme, but it requires you xor the plaintext with the cypher text to get the desired keystream. So the program does not ask for either the plaintext or the cyphertext, but instead them xor-ed together as input.

**Comment by bigfella237** [comment]
These issues should probably be discussed on @N8BWert's GitHub page seeing as they don't really have anything to do with Hashcat.

However, as above, RC4 is a stream cipher. It doesn't encrypt anything per se, it simply generates a predictable stream of data based on the key.

**Comment by funkymonkey5150** [comment]
Has anyone followed this post with any luck for answers

**Comment by poshul** [comment]
If anyone is still interested I spent some time updating an old project I found using CUDA for RC4 generation. The project is here: https://github.com/poshul/rc4_crack_cuda and is super rudimentary right now. But it is fast. I get 300 million guesses per second with my 2070. I'd be happy to have folks submit improvements.

**Comment by bigfella237** [comment]
I see this request just turned eight years old (they grow up so fast)! 

I would still get use out of these modes if anyone is at a loose end and looking for something to do?

**Comment by matrix** [comment]
> I'm guessing there's at least four or five different different "modes" involved depending on whether you can make the key length and/or DropN lengths variable:
> 
> * 40-bit key with no DropN;
> * 40-bit key with a DropN of 256 bytes;
> * 72-bit key with a DropN of 256 bytes;
> * 104-bit key with a DropN of 256 bytes;
> * 104-bit key with a DropN of 267 bytes.

Hi @bigfella237, do you have any test vectors? I don't promise I'll complete the work but I'll try to play with them a little ;)
Thanks

**Comment by matrix** [comment]
A python script output, used to gen the hashes:

```
$ python3 rc4.py 
Key:  6861736863
Plaintext:  48656c6c6f
Ciphertext:  e9a41693b7
hash: $rc4$40$0$e9a41693b7$0$48656c6c6f

Key:  686173686361743132
Plaintext:  48656c6c6f
Ciphertext:  90eaa8d71c
hash: $rc4$72$0$90eaa8d71c$0$48656c6c6f

Key:  68617368636174313233343536
Plaintext:  48656c6c6f
Ciphertext:  a04245c3d7
hash: $rc4$104$0$a04245c3d7$0$48656c6c6f
```

Some tests with new hash-modes (40000, 40001, 40002) and modified inc_cipher_rc4.h/cl:

```
$ ./hashcat -m 40000 '$rc4$40$0$e9a41693b7$0$48656c6c6f' -a3 hashc --potfile-disable -D1 --quiet
$rc4$40$0$e9a41693b7$0$48656c6c6f:hashc
$ ./hashcat -m 40001 '$rc4$72$0$90eaa8d71c$0$48656c6c6f' -a3 hashcat12 --potfile-disable -D1 --quiet
$rc4$72$0$90eaa8d71c$0$48656c6c6f:hashcat12
$ ./hashcat -m 40002 '$rc4$104$0$a04245c3d7$0$48656c6c6f' -a3 hashcat123456 --potfile-disable -D1 --quiet
$rc4$104$0$a04245c3d7$0$48656c6c6f:hashcat123456

```

Details on hash format:

$rc4 => generic rc4 signature
$40 => key bit
$0 => DropN (not implemented yet)
$e9a41693b7 => ciphertext
$0 => plaintext offset (not implemented yet)
$48656c6c6f => plaintext

If you can give me some feedback as to whether the direction is the desired one ... I will try to implement the missing parts.
Thanks ;)

**Comment by matrix** [comment]
implemented dropN

```
$ python3 rc4_dropN.py 
Key:  6861736863
Plaintext:  48656c6c6f
Ciphertext:  e9a41693b7
hash: $rc4$40$0$e9a41693b7$0$48656c6c6f

Key:  6861736863
Plaintext:  48656c6c6f
Ciphertext:  6c72fa10ba
hash: $rc4$40$256$6c72fa10ba$0$48656c6c6f

Key:  686173686361743132
Plaintext:  48656c6c6f
Ciphertext:  90eaa8d71c
hash: $rc4$72$0$90eaa8d71c$0$48656c6c6f

Key:  686173686361743132
Plaintext:  48656c6c6f
Ciphertext:  8c67ebfcd1
hash: $rc4$72$128$8c67ebfcd1$0$48656c6c6f

Key:  68617368636174313233343536
Plaintext:  48656c6c6f
Ciphertext:  a04245c3d7
hash: $rc4$104$0$a04245c3d7$0$48656c6c6f

Key:  68617368636174313233343536
Plaintext:  48656c6c6f
Ciphertext:  7153341e35
hash: $rc4$104$257$7153341e35$0$48656c6c6f

$ ./hashcat -m 40000 '$rc4$40$0$e9a41693b7$0$48656c6c6f' -a3 has?l?l -D1 --quiet --potfile-disable
$rc4$40$0$e9a41693b7$0$48656c6c6f:hashc
$ ./hashcat -m 40000 '$rc4$40$256$6c72fa10ba$0$48656c6c6f' -a3 has?l?l -D1 --quiet --potfile-disable
$rc4$40$256$6c72fa10ba$0$48656c6c6f:hashc
$ ./hashcat -m 40001 '$rc4$72$0$90eaa8d71c$0$48656c6c6f' -a3 hashcat?d?d -D1 --quiet --potfile-disable
$rc4$72$0$90eaa8d71c$0$48656c6c6f:hashcat12
$ ./hashcat -m 40001 '$rc4$72$128$8c67ebfcd1$0$48656c6c6f' -a3 hashcat?d?d -D1 --quiet --potfile-disable
$rc4$72$128$8c67ebfcd1$0$48656c6c6f:hashcat12
$ ./hashcat -m 40002 '$rc4$104$0$a04245c3d7$0$48656c6c6f' -a3 hashcat12?d?d?d?d -D1 --quiet --potfile-disable
$rc4$104$0$a04245c3d7$0$48656c6c6f:hashcat123456
$ ./hashcat -m 40002 '$rc4$104$257$7153341e35$0$48656c6c6f' -a3 hashcat12?d?d?d?d -D1 --quiet --potfile-disable
$rc4$104$257$7153341e35$0$48656c6c6f:hashcat123456
```

**Comment by matrix** [comment]
NVIDIA GeForce RTX 4080, 15736/15981 MB, 76MCU

```
$ for m in $(echo 40000 40001 40002); do ./tools/test.sh -m $m -a all -t all -d1; done
[ test_1746198818 ] > Init test for hash type 40000 (RC4 40-bit DropN).
[ test_1746198818 ] [ Type 40000, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1746198818 ] [ Type 40000, Attack 0, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746198818 ] [ Type 40000, Attack 1, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1746198818 ] [ Type 40000, Attack 1, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746198818 ] [ Type 40000, Attack 3, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/5 not found, 0/5 not matched, 0/5 timeout, 0/5 skipped
[ test_1746198818 ] [ Type 40000, Attack 3, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746198818 ] [ Type 40000, Attack 6, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/3 not found, 0/3 not matched, 0/3 timeout, 0/3 skipped
[ test_1746198818 ] [ Type 40000, Attack 6, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/4 not found, 0/4 not matched, 0/4 timeout, 0/4 skipped
[ test_1746198818 ] [ Type 40000, Attack 7, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1746198818 ] [ Type 40000, Attack 7, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1746198818 ] [ Type 40000, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1746198818 ] [ Type 40000, Attack 0, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746198818 ] [ Type 40000, Attack 1, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1746198818 ] [ Type 40000, Attack 1, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746198818 ] [ Type 40000, Attack 3, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/5 not found, 0/5 not matched, 0/5 timeout, 0/5 skipped
[ test_1746198818 ] [ Type 40000, Attack 3, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746198818 ] [ Type 40000, Attack 6, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/3 not found, 0/3 not matched, 0/3 timeout, 0/3 skipped
[ test_1746198818 ] [ Type 40000, Attack 6, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/4 not found, 0/4 not matched, 0/4 timeout, 0/4 skipped
[ test_1746198818 ] [ Type 40000, Attack 7, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1746198818 ] [ Type 40000, Attack 7, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1746199015 ] > Init test for hash type 40001 (RC4 72-bit DropN).
[ test_1746199015 ] [ Type 40001, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1746199015 ] [ Type 40001, Attack 0, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199015 ] [ Type 40001, Attack 1, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1746199015 ] [ Type 40001, Attack 1, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199015 ] [ Type 40001, Attack 3, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1746199015 ] [ Type 40001, Attack 3, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199015 ] [ Type 40001, Attack 6, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1746199015 ] [ Type 40001, Attack 6, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199015 ] [ Type 40001, Attack 7, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1746199015 ] [ Type 40001, Attack 7, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199015 ] [ Type 40001, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1746199015 ] [ Type 40001, Attack 0, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199015 ] [ Type 40001, Attack 1, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1746199015 ] [ Type 40001, Attack 1, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199015 ] [ Type 40001, Attack 3, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1746199015 ] [ Type 40001, Attack 3, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199015 ] [ Type 40001, Attack 6, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1746199015 ] [ Type 40001, Attack 6, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199015 ] [ Type 40001, Attack 7, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1746199015 ] [ Type 40001, Attack 7, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199397 ] > Init test for hash type 40002 (RC4 104-bit DropN).
[ test_1746199397 ] [ Type 40002, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1746199397 ] [ Type 40002, Attack 0, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199397 ] [ Type 40002, Attack 1, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1746199397 ] [ Type 40002, Attack 1, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199397 ] [ Type 40002, Attack 3, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1746199397 ] [ Type 40002, Attack 3, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199397 ] [ Type 40002, Attack 6, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > Warning : 0/7 not found, 0/7 not matched, 1/7 timeout, 0/7 skipped
[ test_1746199397 ] [ Type 40002, Attack 6, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199397 ] [ Type 40002, Attack 7, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1746199397 ] [ Type 40002, Attack 7, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 1 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199397 ] [ Type 40002, Attack 0, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1746199397 ] [ Type 40002, Attack 0, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199397 ] [ Type 40002, Attack 1, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1746199397 ] [ Type 40002, Attack 1, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199397 ] [ Type 40002, Attack 3, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/8 not found, 0/8 not matched, 0/8 timeout, 0/8 skipped
[ test_1746199397 ] [ Type 40002, Attack 3, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199397 ] [ Type 40002, Attack 6, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > Warning : 0/7 not found, 0/7 not matched, 1/7 timeout, 0/7 skipped
[ test_1746199397 ] [ Type 40002, Attack 6, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
[ test_1746199397 ] [ Type 40002, Attack 7, Mode single, Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/7 not found, 0/7 not matched, 0/7 timeout, 0/7 skipped
[ test_1746199397 ] [ Type 40002, Attack 7, Mode multi,  Device-Type Gpu, Kernel-Type Optimized, Vector-Width 4 ] > OK : 0/1 not found, 0/1 not matched, 0/1 timeout, 0/1 skipped
```

```
$ for m in $(echo 40000 40001 40002); do ./hashcat -b -m $m -d1 --quiet; done
------------------------------------
* Hash-Mode 40000 (RC4 40-bit DropN)
------------------------------------

Speed.#1.........:  2616.0 MH/s (60.49ms) @ Accel:512 Loops:128 Thr:32 Vec:1

------------------------------------
* Hash-Mode 40001 (RC4 72-bit DropN)
------------------------------------

Speed.#1.........:  2570.2 MH/s (61.61ms) @ Accel:128 Loops:512 Thr:32 Vec:1

-------------------------------------
* Hash-Mode 40002 (RC4 104-bit DropN)
-------------------------------------

Speed.#1.........:  2597.1 MH/s (60.93ms) @ Accel:512 Loops:128 Thr:32 Vec:1


```

**Comment by jsteube** [comment]
Renamed 4000x to 3350x
